### PR TITLE
[PRM-795] Updated routing for Upload list page

### PR DESCRIPTION
**Pull request by JBlacklaws**

### CODE REVIEW

File: app/navigation/Navigation.scala [hunk 1]
```diff
@@ -73,7 +75,9 @@ class Navigation @Inject()(dateTimeHelper: DateTimeHelper,
     AppealStartPage -> ((_, _, _) => routes.AppealStartController.onPageLoad()),
     OtherRelevantInformationPage -> ((_, _, _) => routes.OtherReasonController.onPageLoadForUploadEvidence(NormalMode)),
     CancelVATRegistrationPage -> ((answer, _, extraData) => routingForCancelVATRegistrationPage(answer, extraData)),
-    OtherPenaltiesForPeriodPage -> ((_, _, _) => routes.AppealStartController.onPageLoad())
+    OtherPenaltiesForPeriodPage -> ((_, _, _) => routes.AppealStartController.onPageLoad()),
+    UploadFirstDocumentPage -> ((_, _, _) => routes.OtherReasonController.onPageLoadUploadList(NormalMode)),
+    UploadAnotherDocumentPage -> ((_,_,_) => routes.OtherReasonController.onPageLoadUploadList(NormalMode))
```
Comment (LewisMitchell) [inline app/navigation/Navigation.scala:80]: Not that this is an option yet, but you may want to consider adding a page for where to go from the 'You have uploaded X files' i.e. adding UploadListPage and then adding some functionality that would route to the making a late appeal or CYA page when the user says no or to the add another page if the user says yes to adding another document

File: app/navigation/Navigation.scala [hunk 2]
```diff
@@ -48,7 +48,9 @@ class Navigation @Inject()(dateTimeHelper: DateTimeHelper,
     WhyWasTheReturnSubmittedLateAgentPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
     ReasonableExcuseSelectionPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
     WhenDidThePersonDiePage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
-    OtherRelevantInformationPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode))
+    OtherRelevantInformationPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
+    UploadFirstDocumentPage -> ((_, _, _) => routes.OtherReasonController.onPageLoadUploadList(CheckMode)),
+    UploadAnotherDocumentPage -> ((_,_,_) => routes.OtherReasonController.onPageLoadUploadList(CheckMode))
```
Comment (LewisMitchell) [inline app/navigation/Navigation.scala:53]: For CheckMode, if the user selects No - we go straight back to the CYA page.
if the user selects Yes - we go to add another document passing the mode through

File: app/controllers/OtherReasonController.scala [hunk 3]
```diff
@@ -141,6 +145,7 @@ class OtherReasonController @Inject()(whenDidBecomeUnablePage: WhenDidBecomeUnab
       if(request.cookies.get("jsenabled").isDefined) {
         Future(Redirect(controllers.routes.OtherReasonController.onPageLoadForUploadEvidence(mode)))
       } else {
+        val nextPageWithUpload = navigation.nextPage(UploadFirstDocumentPage, mode)
```
Comment (LewisMitchell) [inline app/controllers/OtherReasonController.scala:148]: I don't think you need this, we have logic that will already redirect the user on success and failure

File: app/views/reasonableExcuseJourneys/other/noJs/UploadFirstDocumentPage.scala.html [hunk 4]
```diff
@@ -65,7 +66,7 @@ <h1 class="govuk-heading-l">@messages("otherReason.uploadEvidence.heading")</h1>
             @fileUploadDialog(form)
         </div>
         <p>
-            @govukButton(Button(content = Text(messages("otherReason.uploadEvidence.noJs.button.uploadFile"))))
+            <a href="@nextPageWithUpload" class="govuk-button">@messages("otherReason.uploadEvidence.noJs.button.uploadFile")</a>
```
Comment (LewisMitchell) [inline app/views/reasonableExcuseJourneys/other/noJs/UploadFirstDocumentPage.scala.html:69]: <a> tags do not submit the form. They are hyperlinks that redirects the user to a page without submitting the form itself. You shouldn't need to actually pass anything here and can be removed.

File: app/views/reasonableExcuseJourneys/other/noJs/UploadFirstDocumentPage.scala.html [hunk 5]
```diff
@@ -15,6 +15,7 @@
         fileUploadDialog: components.upload.fileUploadDialog)
 
 @(
+    nextPageWithUpload: String,
```
Comment (LewisMitchell) [inline app/views/reasonableExcuseJourneys/other/noJs/UploadFirstDocumentPage.scala.html:18]: We don't need this as the logic is pre-existing to route the user onward

File: conf/app.routes [hunk 6]
```diff
@@ -92,13 +92,14 @@ GET        /upload-first-document                                   controllers.
...
 GET        /upload-taking-longer-than-expected/change              controllers.OtherReasonController.onPageLoadForUploadTakingLongerThanExpected(mode: Mode = CheckMode)
 POST       /upload-taking-longer-than-expected                     controllers.OtherReasonController.onSubmitForUploadTakingLongerThanExpected(mode: Mode = NormalMode)
 POST       /upload-taking-longer-than-expected/change              controllers.OtherReasonController.onSubmitForUploadTakingLongerThanExpected(mode: Mode = CheckMode)
 GET        /upload-another-document                                controllers.OtherReasonController.onPageLoadForAnotherFileUpload(mode: Mode = NormalMode)
 GET        /upload-another-document/change                         controllers.OtherReasonController.onPageLoadForAnotherFileUpload(mode: Mode = CheckMode)
+GET        /uploaded-documents                                      controllers.OtherReasonController.onPageLoadUploadList(mode: Mode = NormalMode)
+GET        /uploaded-documents/change                               controllers.OtherReasonController.onPageLoadUploadList(mode: Mode = CheckMode)
```
Comment (LewisMitchell) [inline conf/app.routes:102]: This is much better naming than /upload-complete 😊

File: it/controllers/OtherReasonControllerISpec.scala [hunk 7]
```diff
@@ -516,13 +516,13 @@ class OtherReasonControllerISpec extends IntegrationSpecCommonBase {
         (SessionKeys.dateCommunicationSent, "2020-02-08T12:00:00"),
         (SessionKeys.journeyId, "1234")
       )
-      val request = await(controller.onPageLoadForUploadComplete(NormalMode)(fakeRequestWithCorrectKeys))
+      val request = await(controller.onPageLoadUploadList(NormalMode)(fakeRequestWithCorrectKeys))
```
Comment (LewisMitchell) [inline it/controllers/OtherReasonControllerISpec.scala:519]: Could you change the "GET /upload-complete" to the correct URL:
https://github.com/hmrc/penalties-appeals-frontend/pull/185/files#diff-1f25331e29c9f185c275d716035c984e6f80e928d6470edf5c800ab02f647228R508

File: it/controllers/OtherReasonControllerISpec.scala [hunk 8]
```diff
@@ -516,13 +516,13 @@ class OtherReasonControllerISpec extends IntegrationSpecCommonBase {
...
       request.header.status shouldBe Status.OK
     }
 
     "return 303 (SEE_OTHER) when the user is not authorised" in {
       AuthStub.unauthorised()
-      val request = await(buildClientForRequestToApp(uri = "/upload-complete").get())
+      val request = await(buildClientForRequestToApp(uri = "/upload-another-document").get())
```
Comment (LewisMitchell) [inline it/controllers/OtherReasonControllerISpec.scala:525]: This should be "/uploaded-documents"

File: test/controllers/OtherReasonControllerSpec.scala [hunk 9]
```diff
@@ -353,19 +353,19 @@ class OtherReasonControllerSpec extends SpecBase {
 
     "onPageLoadForUploadComplete" should {
```
Comment (LewisMitchell) [inline test/controllers/OtherReasonControllerSpec.scala:354]: Could you change this to "onPageLoadUploadList"

File: test/views/UploadFirstDocumentPageSpec.scala [hunk 10]
```diff
@@ -58,8 +58,9 @@ class UploadFirstDocumentPageSpec extends SpecBase with ViewBehaviours {
       )
     )
     val noUploadAction = controllers.routes.MakingALateAppealController.onPageLoad().url
+    val uploadAction = controllers.routes.OtherReasonController.onPageLoadForAnotherFileUpload(NormalMode).url
```
Comment (LewisMitchell) [inline test/views/UploadFirstDocumentPageSpec.scala:61]: I don't think you need this

File: test/views/UploadFirstDocumentPageSpec.scala [hunk 11]
```diff
@@ -58,8 +58,9 @@ class UploadFirstDocumentPageSpec extends SpecBase with ViewBehaviours {
...
     def applyView(request: FakeRequest[_] = fakeRequest): HtmlFormat.Appendable = {
-      uploadFirstDocumentPage.apply(mockUpscanInitiateResponseModel, form, noUploadAction)(request, implicitly, implicitly)
+      uploadFirstDocumentPage.apply(uploadAction, mockUpscanInitiateResponseModel, form, noUploadAction)(request, implicitly, implicitly)
```
Comment (LewisMitchell) [inline test/views/UploadFirstDocumentPageSpec.scala:63]: As above

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

File: app/navigation/Navigation.scala [hunk 12]
```diff
@@ -48,7 +48,9 @@ class Navigation @Inject()(dateTimeHelper: DateTimeHelper,
     WhyWasTheReturnSubmittedLateAgentPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
     ReasonableExcuseSelectionPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
     WhenDidThePersonDiePage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
-    OtherRelevantInformationPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode))
+    OtherRelevantInformationPage -> ((_, request, _) => routeToMakingALateAppealOrCYAPage(request, CheckMode)),
+    UploadFirstDocumentPage -> ((_, _, _) => routes.OtherReasonController.onPageLoadUploadList(CheckMode)),
+    UploadAnotherDocumentPage -> ((answer, _, _) => routeToAdditionalUploadOrCYAPage(CheckMode, answer))
```
Comment (LewisMitchell) [inline app/navigation/Navigation.scala:53]: I think once the user has uploaded another document and they are in CheckMode, that they should be routed to the upload list page?

Comment (LewisMitchell) [reply app/navigation/Navigation.scala:53]: A case for FileListPage should be added so it then calls  routeForUploadList

File: app/navigation/Navigation.scala [hunk 13]
```diff
@@ -151,4 +156,24 @@ class Navigation @Inject()(dateTimeHelper: DateTimeHelper,
...
+
+  def routeForUploadList(answer: Option[String], request: UserRequest[_], mode: Mode): Call = {
+    answer match {
+      case Some(ans) if ans.equalsIgnoreCase("yes") => routes.OtherReasonController.onPageLoadForAnotherFileUpload(mode)
+      case Some(ans) if ans.equalsIgnoreCase("no") => routeToMakingALateAppealOrCYAPage(request, mode)
+      case _ =>
+        logger.debug("[Navigation][routeForUploadList]: unable to get answer - reloading 'UploadListPage'")
+        routes.OtherReasonController.onPageLoadUploadList(mode)
+    }
+  }
```
Comment (LewisMitchell) [inline app/navigation/Navigation.scala:178]: I don't think you need `routeToAdditionalUploadOrCYAPage` as its practically duplicating `routeForUploadList`

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

**Comment by LewisMitchell** [comment]
Also needs rebasing now

**Comment by LewisMitchell** [comment]
Normal mode - No additional

https://user-images.githubusercontent.com/53828896/140495041-9d14f42f-97cb-4d0c-809f-8b75da98fe06.mov


Normal mode - additional

https://user-images.githubusercontent.com/53828896/140495064-bd51289e-f686-403b-8f80-13d78f483da0.mov


Check mode - No additional

https://user-images.githubusercontent.com/53828896/140495077-e2d03d96-2eca-4ad1-9fea-da0b4da515ee.mov


Check mode - additional

https://user-images.githubusercontent.com/53828896/140495099-1c3963bc-1bc8-49a2-8e86-a96b331149d7.mov


Normal mode - No additional - obligation

https://user-images.githubusercontent.com/53828896/140495191-31cd73bd-752d-47ca-bd35-809144392c63.mov


Normal mode - additional - obligation

https://user-images.githubusercontent.com/53828896/140495219-879f80e1-bcfd-4d0d-ab4f-aee614baf9c3.mov


Check mode - No additional - obligation

https://user-images.githubusercontent.com/53828896/140495243-fc491f1f-bb9d-46d8-9561-fdc45a6dc80f.mov


Check mode - additional - obligation

https://user-images.githubusercontent.com/53828896/140495265-780ca90e-2229-4c09-96de-76e7fe1a9702.mov

**Review by AndrewRButcher — APPROVED** [review]

**Pull request merged**